src/main/java/rx/observables/SyncOnSubscribe.java

```diff
@@ -0,0 +1,292 @@
+
+	}
+
+	/**
+	 * Generates a synchronous onSubscribe function that manages back pressure
+	 * requests.
+	 * 
+	 * @param generator
+	 * @param next
+	 * @param onUnsubscribe
+	 * @return
+	 */
+	@Experimental
+	public static <S, T> SyncOnSubscribe<S, T> create(Func0<S> generator,
+			Func2<? super S, ? super Subscriber<? super T>, ? extends S> next, Action1<? super S> onUnsubscribe) {
+		return new LambdaOnSubscribe<S, T>(generator, next, onUnsubscribe);
+	}
+
+	/**
+	 * @param next
+	 * @return
+	 */
+	@Experimental
+	public static <T> SyncOnSubscribe<Void, T> create(final Action1<Subscriber<? super T>> next) {
+		Func2<Void, Subscriber<? super T>, Void> nextFunc1 = new Func2<Void, Subscriber<? super T>, Void>() {
+			@Override
+			public Void call(Void state, Subscriber<? super T> subscriber) {
+				next.call(subscriber);
+				return state;
+			}
+		};
+		Func2<Void, Subscriber<? super T>, Void> nextFunc = nextFunc1;
+		Action1<Void> onUnsubscribe = new Action1<Void>() {
+			@Override
+			public void call(Void t) {
+
+			}
+		};
+		Func0<Void> generatorFunc = new Func0<Void>() {
+			@Override
+			public Void call() {
+				return null;
+			}
+		};
+		return new LambdaOnSubscribe<Void, T>(generatorFunc, nextFunc, onUnsubscribe);
+	}
+
+	/**
+	 * @param generator
+	 * @param next
+	 * @return
+	 */
+	@Experimental
+	public static <S, T> SyncOnSubscribe<S, T> create(Func0<S> generator,
+			final Action2<S, Subscriber<? super T>> next) {
+		Func2<S, Subscriber<? super T>, S> nextFunc = new Func2<S, Subscriber<? super T>, S>() {
+			@Override
+			public S call(S state, Subscriber<? super T> subscriber) {
+				next.call(state, subscriber);
+				return state;
+			}
+		};
+		Action1<S> onUnsubscribe = new Action1<S>() {
+			@Override
+			public void call(S t) {
+
+			}
+		};
+		return new LambdaOnSubscribe<S, T>(generator, nextFunc, onUnsubscribe);
+	}
+
+	/**
+	 * @param next
+	 * @param onUnsubscribe
+	 * @return
+	 */
+	@Experimental
+	public static <T> SyncOnSubscribe<Void, T> create(final Action1<Subscriber<? super T>> next,
+			final Action0 onUnsubscribe) {
+		Func0<Void> generator = new Func0<Void>() {
+			@Override
+			public Void call() {
+				return null;
+			}
+		};
+		Func2<Void, Subscriber<? super T>, Void> nextFunc = new Func2<Void, Subscriber<? super T>, Void>() {
+			@Override
+			public Void call(Void state, Subscriber<? super T> subscriber) {
+				next.call(subscriber);
+				return null;
+			}
+		};
+		Action1<Void> wrappedOnUnsubscribe = new Action1<Void>() {
+			@Override
+			public void call(Void t) {
+				onUnsubscribe.call();
+			}
+		};
+		return new LambdaOnSubscribe<Void, T>(generator, nextFunc, wrappedOnUnsubscribe);
+	}
+
+	@Override
+	public final void call(final Subscriber<? super T> subscriber) {
+		state = generateState();
+		final SyncOnSubscribe<S, T> self = this;
+		final Subscription s = Subscriptions.create(new Action0() {
+			@Override
+			public void call() {
+				self.onUnsubscribe(state);
```

Comment:
This allows unsubscribing the state while an onNext is in progress which requires extra care for the state object.